terraform/eval_state.go

```diff
@@ -5,16 +5,75 @@ import (
 )
 
 // EvalReadState is an EvalNode implementation that reads the
-// InstanceState for a specific resource out of the state.
+// primary InstanceState for a specific resource out of the state.
 type EvalReadState struct {
-	Name         string
-	Tainted      bool
-	TaintedIndex int
-	Output       **InstanceState
+	Name   string
+	Output **InstanceState
 }
 
-// TODO: test
 func (n *EvalReadState) Eval(ctx EvalContext) (interface{}, error) {
+	return readInstanceFromState(ctx, n.Name, n.Output, func(rs *ResourceState) (*InstanceState, error) {
+		return rs.Primary, nil
+	})
+}
+
+// EvalReadStateTainted is an EvalNode implementation that reads a
+// tainted InstanceState for a specific resource out of the state
+type EvalReadStateTainted struct {
+	Name   string
+	Output **InstanceState
+	// Index indicates which instance in the Tainted list to target, or -1 for the last item.
+	Index int
+}
+
+func (n *EvalReadStateTainted) Eval(ctx EvalContext) (interface{}, error) {
+	return readInstanceFromState(ctx, n.Name, n.Output, func(rs *ResourceState) (*InstanceState, error) {
+		// Get the index. If it is negative, then we get the last one
+		idx := n.Index
+		if idx < 0 {
+			idx = len(rs.Tainted) - 1
+		}
+		if idx >= 0 && idx < len(rs.Tainted) {
+			return rs.Tainted[idx], nil
+		} else {
+			return nil, fmt.Errorf("bad tainted index: %d, for resource: %#v", idx, rs)
+		}
+	})
+}
+
+// EvalReadStateDeposed is an EvalNode implementation that reads the
+// deposed InstanceState for a specific resource out of the state
+type EvalReadStateDeposed struct {
+	Name   string
+	Output **InstanceState
+	// Index indicates which instance in the Deposed list to target, or -1 for the last item.
+	Index int
+}
+
+func (n *EvalReadStateDeposed) Eval(ctx EvalContext) (interface{}, error) {
+	return readInstanceFromState(ctx, n.Name, n.Output, func(rs *ResourceState) (*InstanceState, error) {
+		// Get the index. If it is negative, then we get the last one
+		idx := n.Index
+		if idx < 0 {
+			idx = len(rs.Deposed) - 1
+		}
+		if idx >= 0 && idx < len(rs.Deposed) {
+			return rs.Deposed[idx], nil
+		} else {
+			return nil, fmt.Errorf("bad deposed index: %d, for resource: %#v", idx, rs)
+		}
+	})
+}
+
+// Does the bulk of the work for the various flavors of ReadState eval nodes.
+// Each node just provides a reader function to get from the ResourceState to the
+// InstanceState, and this takes care of all the plumbing.
+func readInstanceFromState(
+	ctx EvalContext,
+	resourceName string,
+	output **InstanceState,
+	reader func(*ResourceState) (*InstanceState, error),
```

Comment:
Go thing, if its a function type, I usually suffix the var with `Fn`, so: `readerFn` in this case. Its up to you but I find that clearer later in the code to realize I'm calling a function from a var vs a function named `reader.